Add FieldConfig to encapsulate a column's encoding and indexing info

siddharthteotia opened the pull request:
TableConfig can take a list of FieldConfig to encapsulate storage, indexing and any other information for a field.

**Motivation:**

As part of [text search feature](https://github.com/apache/incubator-pinot/pull/4993), we need to add a new set of columns to IndexingConfig : something like TextIndexConfig and capture the information required for text indexes: column name and some additional properties used by the implementation.

While this is good and in accordance with the existing model for specifying the per column encoding and indexing info, it is not very clean and interspersed throughout the IndexingConfig.

This PR implements a new model to store the per column encoding and indexing info encapsulated into FieldConfig object. TableConfig can take a list of FieldConfig. A FieldConfig is self-descriptive about the encoding and indexing info for the column along with additional properties.

Please read this document (also shared in the open source community) to know more details, migration plan from existing model to new model etc.

https://docs.google.com/document/d/1PyULfbLOYa5OexZMqbP_MZgg-338flKETJOKWS5en3g/edit

kishoreg commented on pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java:
```diff
@@ -82,7 +88,8 @@ public TableConfig() {
       @Nullable QuotaConfig quotaConfig, @Nullable TableTaskConfig taskConfig, @Nullable RoutingConfig routingConfig,
-      @Nullable Map<InstancePartitionsType, InstanceAssignmentConfig> instanceAssignmentConfigMap) {
+      @Nullable Map<InstancePartitionsType, InstanceAssignmentConfig> instanceAssignmentConfigMap,
+      @Null List<FieldConfig> fieldConfigList) {
```
what is @ Null, is it suppose to be @ Nullable?

siddharthteotia replied:
Yes that was a typo.

siddharthteotia commented:
> make sure there is a test case that can parse a json with missing fieldConfigs

Done

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java:
```diff
@@ -157,7 +168,7 @@ public static TableConfig fromJsonConfig(JsonNode jsonConfig)
       return null;
     }
-    if (childConfigNode.isObject()) {
+    if (childConfigNode.isObject() || childConfigNode instanceof ArrayNode) {
```
Equivalent to `childConfigNode.isContainerNode()`

siddharthteotia replied:
Done

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/config/TableConfig.java:
```diff
@@ -177,7 +188,7 @@ public static TableConfig fromJsonConfig(JsonNode jsonConfig)
       return null;
     }
-    if (childConfigNode.isObject()) {
+    if (childConfigNode.isObject() || childConfigNode instanceof ArrayNode) {
```
Same here

siddharthteotia replied:
Done

Jackie-Jiang commented on pinot-common/src/main/java/org/apache/pinot/common/config/FieldConfig.java:
```diff
@@ -0,0 +1,72 @@
+  private String _name;
+  private EncodingType _encodingType;
+  private IndexType _indexType;
+  private Properties _properties;
```
I don't think we need the functionalities provided by Properties class, maybe Map<String, String> is enough?

siddharthteotia replied:
Yes, Map<String, String> is enough. Properties is more for saving to file.

siddharthteotia commented on pinot-common/src/main/java/org/apache/pinot/common/config/FieldConfig.java:
```diff
@@ -0,0 +1,72 @@
+  private Properties _properties;
+
+  @JsonCreator
+  FieldConfig (@JsonProperty(value = "name", required = true) String name,
```
done

siddharthteotia commented on pinot-common/src/main/java/org/apache/pinot/common/config/FieldConfig.java:
```diff
@@ -0,0 +1,72 @@
+
+
+public class FieldConfig extends BaseJsonConfig {
+  private String _name;
```
Done

siddharthteotia commented on pinot-common/src/main/java/org/apache/pinot/common/config/FieldConfig.java:
```diff
@@ -0,0 +1,72 @@
+    _properties = properties;
+  }
+
+  // If null, we will create raw forward index
```
Yes. Done

siddharthteotia commented:
Addressed all the review comments. 
Circulated the detailed doc in community. There is going to be follow-up work soon to deprecate and migrate -- the information is captured in the doc

https://docs.google.com/document/d/1PyULfbLOYa5OexZMqbP_MZgg-338flKETJOKWS5en3g/edit

siddharthteotia commented:
> LGTM otherwise. Are we going to use this to replace some of the indexing config?

Yes exactly. Some parts of IndexingConfig like invertedIndexColumns, noDictionaryColumns, bloomFilterColumns etc will be gone and replaced by FieldConfig where the column's encoding and index info is self-contained and not spread throughout IndexingConfig.

TableConfig can then have a List<FieldConfig.

Please see here https://docs.google.com/document/d/1PyULfbLOYa5OexZMqbP_MZgg-338flKETJOKWS5en3g/edit
